### PR TITLE
GIMP Plugin: Reduce memory use when loading (and other changes)

### PR DESCRIPTION
**Pull request by xiota**
Convert layers to destination precision as they're decoded, before they're inserted into the image.  Also includes some minor cleanup.  Addresses #638

### CODE REVIEW

File: plugins/gimp/file-jxl-save.cc [hunk 1]
```diff
@@ -789,14 +779,24 @@ bool SaveJpegXlImage(const gint32 image_id, const gint32 drawable_id,
...
+  jxl_save_opts.pixel_format.data_type = JXL_TYPE_FLOAT;
+  jxl_save_opts.SetBablType("float");
+  const Babl* destination_format =
+      babl_format(jxl_save_opts.babl_format_str.c_str());
 
+  // convert precision to float if needed
+  if (jxl_save_opts.basic_info.bits_per_sample < 32) {
+    gimp_image_convert_precision(duplicate, GIMP_PRECISION_FLOAT_LINEAR);
+  } else if (!jxl_save_opts.is_linear) {
+    gimp_image_convert_precision(duplicate, GIMP_PRECISION_FLOAT_LINEAR);
```
Comment (jonsneyers) [inline plugins/gimp/file-jxl-save.cc:798]: `if (jxl_save_opts.basic_info.bits_per_sample < 32 || !jxl_save_opts.is_linear) {` ?

Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:797]: Yes.  Thank you.  Done.

**Review by jonsneyers — APPROVED** [review]

**Review by jonsneyers — CHANGES_REQUESTED** [review]
I think it's better to ignore the original profile and just always use the target profile with float precision. That should simplify things a lot.

**Comment by jonsneyers** [comment]
We don't have CI tests for plugins, but I assume you tried this with various images / encode settings?

There seems to be a merge conflict so please rebase.

**Comment by xiota** [comment]
@jonsneyers  Just rebased.  I tested with the images that were causing problems before.

I noticed when saving with distance = 0 (8-bit images), the jxl files are nearly double the size of those produced by `cjxl`.  Is that because I'm sending float data to the encoder?  (even though basic info is set to 8-bit)

When not *trying* to save ICC profiles, the advanced features become unnecessary, so the save dialog is much simplified...  Problems were... GIMP doesn't accurately report whether the TRC is linear (but the profile description correctly states it).  Color encoding info cannot be changed when an ICC is attached, yet the encoder does not parse the ICC, so the TRC and other values were wrong.  (I think either the encoder needs to parse the ICC to set the correct values or the API should allow the color encoding to be changed, and the ICC is just attached for "informational" purposes.)

I don't know how to check whether lcms2 is working in "unbounded" mode.  The docs state it's capable, but I don't know how to enable it or check whether the conditions are met.  Using lcms2 because GEGL didn't seem to work properly and lcms2 is a GIMP dependency, so *should* be available whenever the plugin is compiled.

screenshot

**Comment by xiota** [comment]
I just tried with a 16-bit image converted from raw.  The file `cjxl` filesize is 91.7M.  The gimp file size is 37.5M  That's about expected since the gimp plugin has `uses_original_profile = false`, which approximately doubles file sizes.

**Comment by jonsneyers** [comment]
Use_original_profile = false is not lossless, and it is much larger because things get converted to XYB internally and then get saved 'losslessly' with about 12 bits per channel (the bits_per_sample does not have influence on what gets encoded when XYB is used). When you want to do lossless, you always want to use the original profile and not convert to XYB.

**Comment by xiota** [comment]
@jonsneyers 

> Use_original_profile = false is not lossless and it is much larger because things get converted to XYB internally and then get saved 'losslessly' with about 12 bits per channel (the bits_per_sample does not have influence on what gets encoded when XYB is used).

The files I've saved with the API when `uses_original_profile = false` are usually smaller (by half) than with `uses_original_profile = true`.

> When you want to do lossless, you always want to use the original profile and not convert to XYB.

The version of the gimp plugin in this PR doesn't save "lossless" (that's why I wrote "d=0").  It can use `d=0` and `uses_original_profile = false`.  The problem with `uses_original_profile` are related to the ICC issues I've been having.

If the encoder won't parse and use the ICC profile, there needs to be a way to set correct color encoding values.  Currently, the API blocks changes to color encoding if an ICC is attached.  And if color encoding is set, an ICC cannot be attached.  Everything works fine if sticking with sRGB, but in that case, there isn't any need to attach a "real" ICC profile.

**Comment by jonsneyers** [comment]
For lossless, the encoder doesn't need to parse the icc profile, since it doesn't need to know the meaning of the pixel data. So I recommend to set the icc profile and uses_original_profile=true in case of lossless, and to convert to linear unclamped sRGB and uses_original_profile=false in case of lossy.

@lvandeve / @sboukortt does that make sense?

**Comment by xiota** [comment]
@novomesk Thank you.  I'm looking at it to see what went wrong...

**Comment by mo271** [comment]
any updates on this pr?

**Comment by mo271** [comment]
Closing for now, please fell to re-open once this cleanly rebases and improves the plugin.

**Pull request closed without merging**